src/blur/meta-blur-actor.c

```diff
@@ -0,0 +1,933 @@
+#include <sys/time.h>
+#include <GL/gl.h>
+#include <clutter/clutter.h>
+
+#include "cogl-utils.h"
+#include "blur-utils.h"
+#include <meta/errors.h>
+#include <meta/screen.h>
+#include <meta/compositor-mutter.h>
+#include <meta-blur-actor.h>
+
+static void _stage_add_always_redraw_actor(ClutterStage *stage, ClutterActor *actor)
+    __attribute__((weakref("clutter_stage_add_always_redraw_actor")));
+static void _stage_remove_always_redraw_actor(ClutterStage *stage, ClutterActor *actor)
+    __attribute__((weakref("clutter_stage_remove_always_redraw_actor")));
+
+enum
+{
+    PROP_META_SCREEN = 1,
+    PROP_WINDOW_ACTOR,
+    PROP_RADIUS,
+    PROP_ROUNDS,
+};
+
+typedef enum {
+    CHANGED_EFFECTS = 1 << 1,
+    CHANGED_SIZE    = 1 << 2,
+    CHANGED_ALL = 0xFFFF
+} ChangedFlags;
+
+
+static void build_gaussian_blur_kernel(int* pradius, float* offset, float* weight)
+{
+    int radius = *pradius;
+    radius += (radius + 1) % 2;
+    int sz = (radius+2)*2-1;
+    int N = sz-1;
+    float sigma = 1.0f;
+
+    float sum = powf(2, N);
+    weight[radius+1] = 1.0;
+    for (int i = 1; i < radius+2; i++) {
+        weight[radius-i+1] = weight[radius-i+2] * (N-i+1) / i;
+    }
+    sum -= (weight[radius+1] + weight[radius]) * 2.0;
+
+    for (int i = 0; i < radius; i++) {
+        offset[i] = (float)i*sigma;
+        weight[i] /= sum;
+    }
+
+    *pradius = radius;
+
+    radius = (radius+1)/2;
+    for (int i = 1; i < radius; i++) {
+        float w = weight[i*2] + weight[i*2-1];
+        float off = (offset[i*2] * weight[i*2] + offset[i*2-1] * weight[i*2-1]) / w;
+        offset[i] = off;
+        weight[i] = w;
+    }
+    
+    *pradius = radius;
+}
+
+struct _MetaBlurActorPrivate
+{
+    guint enabled: 1;
+
+    MetaScreen *screen;
+    MetaWindowActor *window_actor;
+    MetaWindow *window;
+
+    gboolean blurred;
+    int radius;
+    float kernel[101];
+    int rounds;
+
+    ChangedFlags changed;
+    CoglPipeline *template;
+    CoglPipeline *pl_passthrough;
+    CoglPipeline *pl_masked; // masked by shape 
+
+    CoglPipeline *pipeline;
+    CoglPipeline *pipeline2;
+
+    CoglTexture* texture;
+
+    CoglTexture* blur_mask_texture; // for shaped region blur 
+    cairo_surface_t *blur_mask; 
+
+    CoglTexture* fbTex, *fbTex2;
+    CoglOffscreen* fb, *fb2;
+    float fb_width;
+    float fb_height;
+
+    cairo_region_t *clip_region;
+
+    int queued_redraw;
+    int pipeline_res_location;
+    int pipeline2_res_location;
+
+    unsigned int gl_handle;
+    unsigned int gl_target;
+};
+
+G_DEFINE_TYPE (MetaBlurActor, meta_blur_actor,
+        CLUTTER_TYPE_ACTOR);
+
+void (*meta_bind_texture)( GLenum      target,
+            GLuint      texture);
+void (*meta_copy_sub_tex)( GLenum      target, 
+            GLint level, 
+            GLint xoffset, 
+            GLint yoffset,
+            GLint x,
+            GLint y,
+            GLsizei width,
+            GLsizei height);
+
+gboolean meta_blur_actor_get_supported (void)
```

Comment:
This method checks out if the current Cogl context can perform certain actions that are needed to perform the blur effect - if this returns `false` the DBus `enable_blur_behind` method will throw an error.